### PR TITLE
LSTM: Pre-compute x_t * warr

### PR DESCRIPTION
**Pull request by sffc**
Here's another counterintuitive result. I observed that the calculation x_t * warr can always be done ahead of time, and we store one NxHx4 matrix instead of a NxD and a DxHx4 matrix. However, the benchmark shows a major regression on this change, which I don't understand.

### CODE REVIEW

File: experimental/segmenter/src/lstm_bies.rs [hunk 1]
```diff
@@ -75,13 +78,29 @@ impl<'l> Lstm<'l> {
             return Err(Error::DimensionMismatch);
         }
 
+        let mut mat2_precompute = MatrixOwned::new_zero([num_classes, hunits, 4]);
```
Comment (robertbastian) [inline experimental/segmenter/src/lstm_bies.rs:81]: How big are these? Could we precompute them in datagen?

Comment (sffc) [reply experimental/segmenter/src/lstm_bies.rs:81]: Yes I would want to precompute these in datagen if we move forward with this PR, but the benchmarks show a regression that I cannot explain

Comment (sffc) [reply experimental/segmenter/src/lstm_bies.rs:81]: It calls `try_new_*_unstable` outside of the bench function

https://github.com/unicode-org/icu4x/blob/main/experimental/segmenter/benches/bench.rs#L47

Comment (robertbastian) [reply experimental/segmenter/src/lstm_bies.rs:81]: That's odd, where do the 8% that @zbraniecki is seeing on #3210 come from then?

Comment (sffc) [reply experimental/segmenter/src/lstm_bies.rs:81]: The only theory I have is one around memory locality, which helps reduce L1/L2 cache misses. I've seen these calculations be very sensitive to memory locality. I don't really know why this change would hurt memory locality, though. The new matrix is a bit bigger than the old matrix, but it takes the place of 2 matrices.

**Pull request closed without merging**